File: solutions/opus-1/opus/perl/calc.pl
use strict;
use warnings;
use integer;

my (@t, $p);

sub expr {
    my $v = term();
    while ($p < @t && ($t[$p] eq '+' || $t[$p] eq '-')) {
        my $op = $t[$p++];
        my $r = term();
        $v = $op eq '+' ? $v + $r : $v - $r;
    }
    return $v;
}

sub term {
    my $v = factor();
    while ($p < @t && ($t[$p] eq '*' || $t[$p] eq '/')) {
        my $op = $t[$p++];
        my $r = factor();
        if ($op eq '*') {
            $v = $v * $r;
        } else {
            my $q = abs($v) / abs($r);
            $v = (($v < 0) != ($r < 0)) ? -$q : $q;
        }
    }
    return $v;
}

sub factor {
    my $tok = $t[$p++];
    if ($tok eq '(') {
        my $v = expr();
        $p++;
        return $v;
    }
    return 0 + $tok;
}

while (my $line = <STDIN>) {
    @t = $line =~ /(\d+|[-+*\/()])/g;
    next unless @t;
    $p = 0;
    print expr(), "\n";
}

File: solutions/opus-1/opus/ruby/calc.rb
def tdiv(a, b)
  q = a.abs / b.abs
  (a < 0) ^ (b < 0) ? -q : q
end

def evaluate(tokens)
  pos = 0
  factor = nil
  term = nil
  expr = lambda do
    v = term.call
    while pos < tokens.size && (tokens[pos] == '+' || tokens[pos] == '-')
      op = tokens[pos]
      pos += 1
      r = term.call
      v = op == '+' ? v + r : v - r
    end
    v
  end
  term = lambda do
    v = factor.call
    while pos < tokens.size && (tokens[pos] == '*' || tokens[pos] == '/')
      op = tokens[pos]
      pos += 1
      r = factor.call
      v = op == '*' ? v * r : tdiv(v, r)
    end
    v
  end
  factor = lambda do
    t = tokens[pos]
    pos += 1
    if t == '('
      v = expr.call
      pos += 1
      v
    else
      t.to_i
    end
  end
  expr.call
end

out = []
STDIN.each_line do |line|
  tokens = line.scan(/\d+|[-+*\/()]/)
  next if tokens.empty?
  out << evaluate(tokens)
end
puts out
